jerry-main/main-unix-snapshot.c

```diff
@@ -0,0 +1,413 @@
+        }
+      }
+    }
+
+    buffer_p += code_size;
+  }
+  while (buffer_p < buffer_end_p);
+} /* scan_snapshot_functions */
+
+/**
+ * Update all literal offsets in a snapshot data.
+ */
+static void
+update_literal_offsets (uint8_t *buffer_p, /**< snapshot buffer start */
+                        const uint8_t *buffer_end_p, /**< snapshot buffer start */
+                        lit_mem_to_snapshot_id_map_entry_t *lit_map_p) /**< literal map */
+{
+  JERRY_ASSERT (buffer_end_p > buffer_p);
+
+  do
+  {
+    ecma_compiled_code_t *bytecode_p = (ecma_compiled_code_t *) buffer_p;
+    uint32_t code_size = ((uint32_t) bytecode_p->size) << JMEM_ALIGNMENT_LOG;
+
+    if (bytecode_p->status_flags & CBC_CODE_FLAGS_FUNCTION)
+    {
+      jmem_cpointer_t *literal_start_p;
+      uint32_t const_literal_end;
+
+      if (bytecode_p->status_flags & CBC_CODE_FLAGS_UINT16_ARGUMENTS)
+      {
+        literal_start_p = (jmem_cpointer_t *) (buffer_p + sizeof (cbc_uint16_arguments_t));
+
+        cbc_uint16_arguments_t *args_p = (cbc_uint16_arguments_t *) buffer_p;
+        const_literal_end = args_p->const_literal_end;
+      }
+      else
+      {
+        literal_start_p = (jmem_cpointer_t *) (buffer_p + sizeof (cbc_uint8_arguments_t));
+
+        cbc_uint8_arguments_t *args_p = (cbc_uint8_arguments_t *) buffer_p;
+        const_literal_end = args_p->const_literal_end;
+      }
+
+      for (uint32_t i = 0; i < const_literal_end; i++)
+      {
+        if (literal_start_p[i] != JMEM_CP_NULL)
+        {
+          lit_mem_to_snapshot_id_map_entry_t *current_p = lit_map_p;
+
+          while (current_p->literal_id != literal_start_p[i])
+          {
+            current_p++;
+          }
+
+          literal_start_p[i] = current_p->literal_offset;
+        }
+      }
+    }
+
+    buffer_p += code_size;
+  }
+  while (buffer_p < buffer_end_p);
+} /* update_literal_offsets */
+
+/**
+ * Print help.
+ */
+static void
+print_help (char *name) /**< program name */
+{
+  printf ("Usage: %s [COMMAND] [ARGUMENTS]\n"
+          "\n"
+          "Command list:\n"
+          "  merge [OUTPUT_FILE] [INPUT_FILEs]...\n"
+          "      Merge input files into output file.\n"
+          "  help\n"
+          "      Prints this help.\n"
+          "\n",
+          name);
+} /* print_help */
+
+/**
+ * Main function.
+ *
+ * @return error code (0 - no error)
+ */
+int
+main (int argc, /**< number of arguments */
+      char **argv) /**< argument list */
+{
+  if (argc <= 1 || (argc == 2 && (!strcmp ("help", argv[1]))))
+  {
+    print_help (argv[0]);
+    return JERRY_STANDALONE_EXIT_CODE_OK;
+  }
+
+  if (strcmp ("merge", argv[1]))
+  {
+    printf ("Unknown command: %s\n", argv[1]);
+    print_help (argv[0]);
+    return JERRY_STANDALONE_EXIT_CODE_OK;
+  }
+
+  if (argc < 5)
+  {
+    printf ("At least an output and two input files must be passed\n");
+    return JERRY_STANDALONE_EXIT_CODE_FAIL;
+  }
```

Comment:
If this tool is indeed going to be a multi-purpose tool then sub-commands are justified. I've added a commit to #1809 that adds sub-command handling support to the CLI code.

Here is a completely working code that does all what seems to be needed by the current variant of this tool (should work almost by copy-pasting):

```c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include "cli.h"

/* Exit code */
#define JERRY_STANDALONE_EXIT_CODE_OK   (0)
#define JERRY_STANDALONE_EXIT_CODE_FAIL (1)

/** General options */
typedef enum
{
  OPT_HELP,
  OPT__FIRST_NON_GENERAL /**< sub-command-specific options should start from here */
} snap_opt_id_t;

static cli_opt_t snap_opts[] =
{
  CLI_OPT_DEF (.id = OPT_HELP, .opt = "-h", .longopt = "--help", .quant = CLI_QUANT_1,
               .help = "print this help and exit"),
  CLI_OPT_DEF (.id = CLI_OPT_END)
};

/** Options specific to the merge command */
typedef enum
{
  OPT_MERGE_OUT = OPT__FIRST_NON_GENERAL
} merge_opt_id_t;

static cli_opt_t merge_opts[] =
{
  CLI_OPT_DEF (.id = OPT_HELP, .opt = "-h", .longopt = "--help",
               .help = "print this help and exit"),
  CLI_OPT_DEF (.id = OPT_MERGE_OUT, .opt = "-o", .argc = 1, .meta = "FILE", .quant = CLI_QUANT_1,
               .help = "output merged snapshot file (mandatory)"),
  CLI_OPT_DEF (.id = CLI_OPT_POSITIONAL, .meta = "FILE", .quant = CLI_QUANT_P,
               .help = "input snapshot files"),
  CLI_OPT_DEF (.id = CLI_OPT_END)
};

/** Tool sub-commands */
typedef enum
{
  CMD_MERGE
} snap_cmd_id_t;

static cli_cmd_t snap_cmds[] =
{
  CLI_CMD_DEF (.id = CMD_MERGE, .cmd = "merge", .opts = merge_opts,
               .help = "merge snapshot file"),
  CLI_CMD_DEF (.id = CLI_CMD_END)
};

/** argv[0] */
static char *progname;

/**
 * Check whether a usage-related condition holds. If not, print an error
 * message, print the usage, and terminate the application.
 */
static void
check_usage (bool condition, /**< the condition that must hold */
             const char *name, /**< name of the application/command */
             cli_opt_t *cli_opts, /**< command line options to print usage for */
             const char *msg, /**< error message to print if condition does not hold */
             const char *opt) /**< optional part of the error message */
{
  if (!condition)
  {
    printf /*jerry_port_log*/ (/*JERRY_LOG_LEVEL_ERROR, */"%s%s\n", msg, opt != NULL ? opt : "");
    cli_opt_usage (name, cli_opts);
    exit (JERRY_STANDALONE_EXIT_CODE_FAIL);
  }
} /* check_usage */

/** The function that deals with the merge sub-command */
int
do_merge (int argc, /**< number of command line arguments (beyond "argv[0] merge") */
          char **argv) /**< array of command line arguments (beyond "argv[0] merge") */
{
  CLI_CMD_NAME (cmdname, progname, "merge");

  const char *in_files[argc];
  int in_file_counter = 0;
  const char *out_file = NULL;

  cli_opt_state_t state = cli_opt_init (merge_opts, argc, argv);
  for (bool process = true; process;)
  {
    switch (cli_opt_process (&state))
    {
      case OPT_HELP:
      {
        cli_opt_usage (cmdname, merge_opts);
        printf ("\n");
        cli_opt_help (merge_opts);
        return JERRY_STANDALONE_EXIT_CODE_OK;
      }
      case OPT_MERGE_OUT:
      {
        check_usage (out_file == NULL, cmdname, merge_opts, "Error: output file name already specified", NULL);
        out_file = state.arg[1];
        break;
      }
      case CLI_OPT_POSITIONAL:
      {
        in_files[in_file_counter++] = state.arg[0];
        break;
      }
      case CLI_OPT_END:
      {
        process = false;
        break;
      }
      case CLI_OPT_INCOMPLETE:
      {
        check_usage (false, cmdname, merge_opts, "Error: incomplete option: ", state.arg[0]);
        break;
      }
      case CLI_OPT_UNKNOWN:
      default:
      {
        check_usage (false, cmdname, merge_opts, "Error: unrecognized option: ", state.arg[0]);
        break;
      }
    }
  }

  check_usage (in_file_counter > 1, cmdname, merge_opts, "Error: at least two input files must be specified", NULL);
  check_usage (out_file != NULL, cmdname, merge_opts, "Error: output file name must be specified", NULL);

  // below this point, you have in_files[], in_file_counter, out_file set up
  // and you can write the file names in any order, you wont have to keep it in mind
  //    jerry-snap merge -o outf inf1 inf2
  //    jerry-snap merge inf1 inf2 -o outf
  // (and not overwrite something by mistake, -o will always mark the output explicitly)
  // and if you do, you still have a nice CLI with usual -h, --help

  /* THIS IS DUMMY CODE */
  printf ("in files:");
  for (int i = 0; i < in_file_counter; i++)
  {
    printf (" %s", in_files[i]);
  }
  printf ("\n");
  printf ("out file: %s\n", out_file);

  return JERRY_STANDALONE_EXIT_CODE_OK;
} /* do_merge */

/** The function that deals with the case when no sub-command was specified. */
int
do_none (int argc, /**< number of command line arguments (beyond "argv[0]") */
         char **argv) /**< array of command line arguments (beyond "argv[0]") */
{
  cli_opt_state_t state = cli_opt_init (merge_opts, argc, argv);
  for (bool process = true; process;)
  {
    switch (cli_opt_process (&state))
    {
      case OPT_HELP:
      {
        cli_cmd_usage (progname, snap_cmds);
        cli_opt_usage (progname, snap_opts);
        printf ("\n");

        printf ("commands:\n");
        cli_cmd_help (snap_cmds);
        printf ("\n");

        printf ("general options:\n");
        cli_opt_help (snap_opts);
        printf ("\n");

        for (cli_cmd_t *cmds = snap_cmds; cmds->id != CLI_CMD_END; cmds++)
        {
          printf ("%s options:\n", cmds->cmd);
          cli_opt_help (cmds->opts);
        }

        return JERRY_STANDALONE_EXIT_CODE_OK;
      }
      case CLI_OPT_END:
      {
        process = false;
        break;
      }
      case CLI_OPT_POSITIONAL:
      case CLI_OPT_INCOMPLETE:
      case CLI_OPT_UNKNOWN:
      default:
      {
        check_usage (false, progname, snap_opts, "Error: unrecognized option: ", state.arg[0]);
        break;
      }
    }
  }

  check_usage (false, progname, snap_opts, "Error: empty command line", NULL);
  return JERRY_STANDALONE_EXIT_CODE_FAIL;
} /* do_none */

/** Tool entry point: a dispatcher for sub-commands */
int
main (int argc, /**< number of command line arguments, including argv[0] */
      char **argv) /**< array of command line arguments, including argv[0] */
{
  progname = argv[0];
  argc--;
  argv++;

  cli_cmd_state_t state = cli_cmd_init (snap_cmds, argc, argv);
  switch (cli_cmd_process (&state))
  {
    case CMD_MERGE:
    {
      return do_merge (argc - 1, argv + 1);
    }
    case CLI_CMD_NONE:
    {
      return do_none (argc, argv);
    }
    case CLI_CMD_UNKNOWN:
    default:
    {
      check_usage (false, progname, snap_opts, "Error: unrecognized command: ", state.arg[0]);
      break;
    }
  }

  return JERRY_STANDALONE_EXIT_CODE_FAIL;
} /* main */
```

Some example runs:
```
$ gcc -o dummy-snap dummy-snap.c cli.c

$ ./dummy-snap
Error: empty command line
./dummy-snap -h

$ ./dummy-snap -h
./dummy-snap merge [-h] -o FILE FILE...
./dummy-snap -h

commands:
  merge                 merge snapshot file

general options:
  -h, --help            print this help and exit

merge options:
  -h, --help            print this help and exit
  -o FILE               output merged snapshot file (mandatory)
  FILE                  input snapshot files

$ ./dummy-snap -x
Error: unrecognized option: -x
./dummy-snap -h

$ ./dummy-snap x
Error: unrecognized command: x
./dummy-snap -h

$ ./dummy-snap merge
Error: at least two input files must be specified
./dummy-snap merge [-h] -o FILE FILE...

$ ./dummy-snap merge -h
./dummy-snap merge [-h] -o FILE FILE...

  -h, --help            print this help and exit
  -o FILE               output merged snapshot file (mandatory)
  FILE                  input snapshot files

$ ./dummy-snap merge -o outfile infile1 infile2
in files: infile1 infile2
out file: outfile

$ ./dummy-snap merge infile1 infile2 -o outfile
in files: infile1 infile2
out file: outfile

$ ./dummy-snap merge -x
Error: unrecognized option: -x
./dummy-snap merge [-h] -o FILE FILE...
```